### PR TITLE
Do transaction

### PR DESCRIPTION
**Pull request by nagua**

### CODE REVIEW

**Comment by JanPlitzkow** [comment]
I would give an additional message of success after completed transaction...

**Comment by JanPlitzkow** [comment]
And maybe do the show_account_data, too...

File: abrechnung/abrechnungsbot.py [hunk 1]
```diff
@@ -80,10 +82,36 @@ def add_event(self, bot, update, args):
...
+    """
+    group_id = update.message.chat_id
+    logger = logging.getLogger('do_transaction')
+    logger.info("Start do transaction")
+
+    if len(args) != 3:
+      bot.sendMessage(chat_id=group_id, text="This command requires three arguments")
+      return
+
+    amount = args[0]
```
Comment (HWoidt) [inline abrechnung/abrechnungsbot.py]: amount, source, destination = args

File: abrechnung/abrechnungsbot.py [hunk 2]
```diff
@@ -80,10 +82,36 @@ def add_event(self, bot, update, args):
...
+    source = args[1]
+    destination = args[2]
+
+    try:
+      self.groups[group_id].do_transaction(t.Transaction(amount, source, destination))
```
Comment (HWoidt) [inline abrechnung/abrechnungsbot.py]: t.Transaction(*args)

File: abrechnung/group.py [hunk 3]
```diff
@@ -29,44 +30,60 @@ def add_event(self, event):
...
+    add_event_text = "[add_event] - cost_per_person: {cost_per_person}, remainder: {remainder}"
+    print(event)
+    print(add_event_text.format(**{'cost_per_person': cost_per_person, 'remainder': remainder}))
+
+    self.events.append(event)
+
+  def do_transaction(self, transaction):
+    """Transfer money from one persion to another"""
+    found_source, found_destination = False, False
+    for acc in self.accounts:
```
Comment (HWoidt) [inline abrechnung/group.py]: why not:
```python
src, dst = None, None
for acc in self.accounts:
    if acc.name.lower() == transaction.source.lower():
        src = acc
    if acc.name.lower() == transaction.source.lower():
        dst = acc

if not src or nor dst:
    raise GroupError("...")

print(transaction)
src.balance -= transaction.amount_in_cents
dst.balance += transaction.amount_in_cents
self.transactions.append(transaction)
```

File: abrechnung/transaction.py [hunk 4]
```diff
@@ -1,12 +1,15 @@
...
+import time
+
 class Transaction:
 
-  def __init__(self, amount, from_user, to_user):
-    self.amount_in_cents = amount
-    self.from_user = from_user
-    self.to_user = to_user
+  def __init__(self, amount_in_euro, source, destination):
+    self.amount_in_cents = int(float(amount_in_euro) * 100)
```
Comment (HWoidt) [inline abrechnung/transaction.py]: Is correct rounding needed here?
```python
self.amount_in_cents = round(float(amount_in_euro) * 100)
```

**Comment by HWoidt** [comment]
Another suggestion, as there are several occasions of walking lists and comparing values...
```python
class NormalizingDict:
    def __init__(self, items=[]):
        self.data = {}
        for k,v in items:
            self.data[self.normalize(k)] = (k,v)
    def __setitem__(self, key, value):
        self.data[self.normalize(key)] = (key,value)
    def __getitem__(self, key):
        return self.data[self.normalize(key)][1]
    def __contains__(self, key):
        return self.normalize(key) in self.data
    def normalize(self, key):
        return key.lower()
```

File: abrechnung/transaction.py (hunk 4 above)
Comment (nagua) [inline abrechnung/transaction.py]: If it is needed there, we also need it in Event constructor.

**Comment by nagua** [comment]
This is interesting. But once the goal was to use some kind of distance function (levenshtein). This would not help then, would it?

**Comment by HWoidt** [comment]
yes and no. You could modify this somewhat:
```python
class FuzzyDict:
    def __init__(self, items=[]):
        self.data = dict(items)
    def __setitem__(self, key, value):
        similarity, match = self.closest(key)
        if similarity>=self.set_thresh:
             self.data[match] = value
        else:
             raise KeyError()
    def __getitem__(self, key):
        similarity, match self.closest(key)
        if similarity>self.get_thresh:
             return self.data[match]
        else:
              raise KeyError()
    def __contains__(self, key):
        try:
           self[key]
           return True
        except KeyError as e:
           return False
    def closest(self, key):
        # dist is the distance function
        return min((dist(match, key), match) for match in self.data)
```
question is whether addition would be allowed in __setitem__

**Pull request merged**